packages/edit-post/src/components/layout/index.js

```diff
@@ -69,6 +70,7 @@ function Layout() {
 		previousShortcut,
 		nextShortcut,
 		hasBlockSelected,
+		isEntitiesSavedStatesOpen,
```

Comment:
We already have a "pre publish panel" on edit-post/editor, I feel this should just use the same selectors/controls. Or even render `EntitiesSavedStates` on the existing PrePublishPanel instead of having a separate panel.

Reply:
Im not sure I agree.

The pre-publish panel is only currently used in the pre-published state of the post editor.  The multi entity saving panel will need to be used regardless of whether or not the post is published.  If we wanted to use the same selectors/controls or the same panel, we would need to do some refactoring of the pre-publish panel itself to play nice and make sense in both pre and post publish modes.  As things sit right now, the separate panel both plays nicely with the existing pre-publish panel, allows for separation of concerns (entity saving vs. publishing), and keeps a parity of interface between multi entity saving in the site and post editors.

Reply:
> The pre-publish panel is only currently used in the pre-published state of the post editor

For the user, it's just the same pre-publish panel filled with different UIs depending on the state of its current changes.

> If we wanted to use the same selectors/controls or the same panel, we would need to do some refactoring of the pre-publish panel itself to play nice and make sense in both pre and post publish modes

Would that be an issue, I mean why is that not a good approach? 

>  As things sit right now, the separate panel both plays nicely with the existing pre-publish panel, allows for separation of concerns (entity saving vs. publishing)

I don't think this separation as a necessary one, for me, the user just adds and edits things in its content, if it happens to be things saved to multiple entities (like adding a site title block inside its post content), the pre publish panel adapts to clarify what is being saved. That said, I'd be happy to have more design thoughts here @jasmussen @MichaelArestad

Reply:
I'm very late to this one, and honestly I missed the part where #20421 landed on a sidebar solution. Which I personally feel like a sidebar is a wrong solution, exactly because it mimics the pre-publish flow as to be mistaken for it, when it is in fact quite different. 

The pre-publish flow is designed around providing a last minute checkup to _publishing new pages and posts_: surfacing for you things like tags or the permalink in case you dismissed the sidebar when you wrote. It even surfaces once more as a sanity check things like Jetpack publicize to let you know one last time before you publish that "hey — I'm about to ping every social network you have connected". It's an "are you sure you want to publish this post or page"?

Saving multiple template parts, also needs a pre-publish check but for a different reason that has nothing to do with tags, or pinging social networks. You may have inadvertently made changes to the entire website, so you need to be sure that's what you actually want to do. And because saving multiple templates at once is inherently a much more complex task than publishing a post, and making the two look the same does not feel like the right solution. For example the pre-publish flow has been optimized to allow you to "double-click" the publish button to _go fast_. But when you're saving multiple template parts, you really don't want to go fast.

So to the questions at hand there are two opposing views:

> As things sit right now, the separate panel both plays nicely with the existing pre-publish panel, allows for separation of concerns (entity saving vs. publishing)

.. and:

> I don't think this separation as a necessary one, for me, the user just adds and edits things in its content, if it happens to be things saved to multiple entities (like adding a site title block inside its post content), the pre publish panel adapts to clarify what is being saved.

Based on the premise that I think the foundation is wrong, it's hard to pick either. In fact this very stacking of sidebars on top of each other feels like an example of the kind of confusion of different saving contexts, that I believe we should avoid. 

However given a consensus was reached #20421, I will simply note this disagreement and suggest that of the two approaches, the one that at least keeps the panels different, is the one that I would go with, even if it means stacking sidebars.

Also worth noting, going that route, it means you'll need to add an additional on-focus tab stop for screen reader users to the one we already have:

![tabbing](https://user-images.githubusercontent.com/1204802/79461952-27d2d680-7ff7-11ea-83a9-aae6e67d3718.gif)

Reply:
> Would that be an issue, I mean why is that not a good approach?

Refactoring it to play nice isnt an issue, but the reasoning is a bit confusing.  We have a pre-publish panel designed to be used in the pre-published state... updating it to also work in a state where its underlying purpose is no longer relevant seems odd.  Especially when the motivation to do so is to add functionality that is not necessarily related to publishing.

And kind of going with your comment from earlier:
> we shouldn't be afraid of duplicating code. Sometimes I feel early abstractions do more harm than good.

Publishing and entity saving do function separately, and it does feel like trying to couple them together in one editor, while having a separate panel for the other, could unnecessarily make things more complicated and things more difficult for others to work with down the road?

Im not against doing the work to make it happen, just a bit concerned and confused about the reasoning.  

I'd like to hear what design thinks about sharing entity saving and publishing options in the same panel in the pre-publish state.  It seems like it could really clutter the panel as we converge on implementing more designs and functionalities as discussed in #20421, especially when there are a handful of entities in use.

Reply:
>  it means you'll need to add an additional on-focus tab stop for screen reader users to the one we already have:

Im not sure I follow.  The 'one we already have' as in the pre-publish panel doesn't work with screen readers currently?  Or something about adding this extra panel breaks that functionality?  I'm not super experienced with screen readers so I don't completely understand what is needed.

From a tab-stop accessibility perspective, it seems like both panels still work well.  The only conflicting issue I see is when the entity saving panel is open, you can tab down to 'open the publish panel', which will open the publish panel hidden behind the saving panel.  Saving through the save panel triggers the publish panel to open right afterwards, so maybe it would be best to disable/hide that button when the saving panel is open?  The other thought is to have it close the save panel to allow viewing of the publish panel, but then you are in the publish panel with unsaved entities which doesn't really make sense.  🤔

Reply:
Maybe that's one of those areas where it's hard to agree on something.

Noting here that it's possible to do both things at the same time "publish a post" and "save multiple entities", for that reason I find it confusing to offer multiple UIs to the user, I'd prefer some kind of two steps wizard but in the same UI. 

Now technically, the current publish panel already has the required a11y work to make it work ( a hidden button that shows up when you tab through the UI and the corresponding area-region), using a different panel, means another aria-region, another hidden button to open that specific panel. I feel that's wrong, it's like we're continuing to add new panels to the UI, they are visible to screenreader users. Imagine if they were visible to all users having and you have to see all these things at the same time? I also fear the bugs this might produce, all the logic about which panel to show/hide can become complex.

Reply:
Those are some good points @youknowriad.

With the current flow:  when there are multiple entities requiring saving, the 'Publish' button opens the entities saved states (either modal or sidebar) instead of the pre-publish panel.  On save, the supplied callback function then triggers the publish panel to open.  At this point, we only have 1 panel open at a time.  

However, if we do press the hidden button you mention, it opens the publish panel regardless of whether or not there are other entities requiring saving.  From here if we try to 'publish' it will open the save panel on top of the publish panel (if it wasnt already open when pressing the hidden button).  The most simple solution I can think of at the moment would be to make that hidden button behave the same way as the 'publish' button at the top (shouldn't it be doing that anyways?) and open the save panel first if necessary.  This way we should never have both sidebars open at the same time, I think the logic would be fairly simple and essentially mirror what is already going on within the standard 'publish' button. 🤔

Reply:
I am starting to see some advantages with sharing the same panel.  However I do wonder:

> I also fear the bugs this might produce, all the logic about which panel to show/hide can become complex.

If we were sharing the same panel, wouldn't the logic about which content to show/hide within the panel be just as complex?

Reply:
> If we were sharing the same panel, wouldn't the logic about which content to show/hide within the panel be just as complex?

I think it will be simpler personally because of the fact that these two panels show at the same time exactly if you see it this way:

 1- Entities is the first step in the pre-publish wizard
 2- Current pre-publish is the second one

the panel just opens when you click the button on the toolbar (or the hidden button)
The panel closes when you finish the saving

That said, that's the theory, may I propose that if we do a refactor, we do it as a separate branch/PR to compare?

Reply:
> That said, that's the theory, may I propose that if we do a refactor, we do it as a separate branch/PR to compare?

Certainly, I was thinking the same.  Il branch off of this one to explore refactoring around sharing the panel.

Reply:
@youknowriad  - After looking into this a bit more, I think the below solution will make the most sense (commit 1f52b14).  This change to the actions section of the Layout component will ensure that only 1 panel is rendered at any given time, so it should solve our concern about overlapping panels and screen readers with very simple logic.  This allows all behavior to remain the same without rendering any overlapping components.

I thought of the above while looking into refactoring the post-publish-panel to share the saving functionality.  Doing the latter seems to be a fairly complex refactor with much more room for errors in logic for desired functionality.  I don't think it makes sense given how we are currently set up, especially if the above solution will solve our concerns.

<hr/>

I have some lingering questions/concerns regarding the hidden 'Open publish panel' button given questionable behavior **after the post is in the published state** (even on the current master branch w/o the experiment enabled), maybe @jasmussen would have some insight here:
* After a post is in the published state, this hidden 'Open publish panel' button is still present.  Should it be?  There does not seem to be another way to open the panel in this state, so I'd assume the 'hidden' accessibility button shouldn't be present either?
* If it is supposed to sill be present at this stage:  It only actually opens the panel if the post is not dirtied.  If selected after changes have been made, it does not open the panel when clicked.  I assume this is a bug, or is there something else I am missing?

I assume we want this style of 'hidden' button to open the save panel when applicable both in edit-post as well as edit-site?  I could just use some confirmation about how the current implementation is supposed to work though, so I can set things up with the desired functionality in the post editor.